mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt

```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
+    ) = addEffectHandler<G>(cancellationPolicy) { effect, _ ->
         consumer.invoke(effect)
-        flowOf()
     }
 
     inline fun <reified G : F> addProducer(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
         crossinline producer: suspend () -> E
-    ) = addFlowProducer<G> {
+    ) = addEffectHandler<G>(cancellationPolicy) { _, eventsChannel ->
         val event = producer.invoke()
-        flowOf(event)
+        eventsChannel.send(event)
     }
 
     inline fun <reified G : F> addFunction(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
         crossinline function: suspend (G) -> E
-    ): CoroutinesSubtypeEffectHandlerBuilder<F, E> = addFlowProducer<G> { effect ->
+    ) = addEffectHandler<G>(cancellationPolicy) { effect, eventsChannel ->
         val event = function.invoke(effect)
-        flowOf(event)
+        eventsChannel.send(event)
+    }
+
+    inline fun <reified G : F> addFlow(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
+        crossinline flowCollectorFunction: suspend FlowCollector<E>.(G) -> Unit
+    ) = addEffectHandler<G>(cancellationPolicy) { effect, eventsChannel ->
+        flow { flowCollectorFunction(effect) }
+            .collect { event -> eventsChannel.send(event) }
     }
 
     inline fun <reified G : F> addFlowProducer(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
         crossinline function: suspend (G) -> Flow<E>
+    ) = addEffectHandler<G>(cancellationPolicy) { effect, eventsChannel ->
+        function.invoke(effect)
+            .collect { event -> eventsChannel.send(event) }
+    }
+
+    inline fun <reified G : F> addEffectHandler(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
+        crossinline function: suspend (G, SendChannel<E>) -> Unit
     ): CoroutinesSubtypeEffectHandlerBuilder<F, E> {
-        addEffectHandler(G::class) { effect ->
-            function.invoke(effect as G)
-        }
+        addEffectHandler(G::class, when (cancellationPolicy) {
+
+            RunSequentially -> { effectChannel, eventsChannel ->
+                for (effect in effectChannel) {
+                    function.invoke(effect as G, eventsChannel)
+                }
+            }
+
+            RunConcurrently -> { effectChannel, eventsChannel ->
+                coroutineScope {
+                    for (effect in effectChannel) {
+                        launch { function.invoke(effect as G, eventsChannel) }
+                    }
+                }
+            }
+
+            CancelPrevious -> { effectChannel, eventsChannel ->
+                coroutineScope {
+                    var currentJob: Job = Job()
+                    for (effect in effectChannel) {
+                        currentJob.cancel()
+                        currentJob = launch { function.invoke(effect as G, eventsChannel) }
+                    }
+                }
+            }
```

Comment:
If this code would be embedded in the respective strategy objects instead of implemented here, I think that would open up for multiple/custom implementations of the policy. 

So, basically, something like `executionPolicy.create(effectChannel, eventsChannel).run(function)`, maybe?

Reply:
Great suggestion. I changed `ExecutionPolicy` to be an `interface` instead of a `sealed interface`. That gives more flexibility when using `CoroutinesSubtypeEffectHandlerBuilder` "add..." functions.